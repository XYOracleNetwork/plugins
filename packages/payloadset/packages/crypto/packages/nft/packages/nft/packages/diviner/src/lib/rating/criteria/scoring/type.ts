import { NftInfoFields } from '@xyo-network/crypto-nft-payload-plugin'
import { FAIL, PASS, PassFailScoringFunction } from '@xyo-network/crypto-nft-score-model'

export const scoreType: PassFailScoringFunction<NftInfoFields> = (nft: NftInfoFields) => {
  if (!nft.type) return FAIL
  if (typeof nft.type !== 'string') return FAIL
  const type = nft.type.toUpperCase()
  if (type !== 'ERC721' && type !== 'ERC1155') return FAIL
  return PASS
}
