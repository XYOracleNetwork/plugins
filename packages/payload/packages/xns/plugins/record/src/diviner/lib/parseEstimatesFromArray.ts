import { exists } from '@xylabs/exists'
import { type BoundWitness, isBoundWitness } from '@xyo-network/boundwitness-model'
import {
  type HashLeaseEstimate, HashLeaseEstimateSchema,
  isHashLeaseEstimateWithMeta,
} from '@xyo-network/diviner-hash-lease'
import { PayloadBuilder } from '@xyo-network/payload-builder'
import type {
  Payload, WithMeta, WithSources,
} from '@xyo-network/payload-model'

import {
  type DomainRegistrationLease, DomainRegistrationLeaseSchema, isDomainRegistrationLeaseWithMeta,
} from '../../DomainRegistration/index.ts'

export type Estimate = [BoundWitness, WithSources<HashLeaseEstimate>, DomainRegistrationLease]

/**
 * Parses a generic array of payloads into estimates
 * @param payloads An array of payloads
 * @returns Parsed estimates
 */
export const parseEstimatesFromArray = async (payloads?: Payload[]): Promise<Estimate[]> => {
  if (!payloads) return []
  const bws = payloads
    // Find all bound witnesses
    .filter(isBoundWitness)
    // that have the required estimate payloads
    .filter(hasEstimatePayloads) ?? []
  if (bws.length === 0) return []
  const hashMap = await PayloadBuilder.toHashMap(payloads)
  return bws.map<Estimate | undefined>((bw) => {
    const hashLeaseEstimate = getPayloadBySchemaFromBoundWitness<WithSources<HashLeaseEstimate>>(
      bw,
      HashLeaseEstimateSchema,
      hashMap,
      isHashLeaseEstimateWithMeta,
    )
    if (!hashLeaseEstimate) return
    const domainLease = getPayloadBySchemaFromBoundWitness<DomainRegistrationLease>(
      bw,
      DomainRegistrationLeaseSchema,
      hashMap,
      isDomainRegistrationLeaseWithMeta,
    )
    if (!domainLease) return
    return [bw, hashLeaseEstimate, domainLease]
  }).filter(exists)
}

/**
 * Tests if a BoundWitness has the required estimate payloads
 * @param bw The BoundWitness to check for estimate payloads
 * @returns True if the BoundWitness has the required estimate payloads, false otherwise
 */
const hasEstimatePayloads = (bw: BoundWitness): boolean => {
  return bw.payload_schemas.includes(DomainRegistrationLeaseSchema)
    && bw.payload_schemas.includes(HashLeaseEstimateSchema)
}

/**
 * Finds the first instance of a payload in a BoundWitness by schema
 * @param bw The BoundWitness to extract the payload from
 * @param schema The schema of the payload to extract
 * @param hashMap The hashmap of the payloads
 * @param identity An identity function to check if the payload is of the correct type
 * @returns The payload if it exists or undefined
 */
const getPayloadBySchemaFromBoundWitness = <T extends Payload = Payload>(
  bw: BoundWitness,
  schema: string,
  hashMap: Awaited<ReturnType<typeof PayloadBuilder.toHashMap>>,
  identity: (payload: Payload) => payload is WithMeta<T>,
): WithMeta<T> | undefined => {
  const schemaIndex = bw.payload_schemas.indexOf(schema)
  if (schemaIndex === -1) return
  const hash = bw.payload_hashes[schemaIndex]
  return [hashMap[hash]].filter(exists).find(identity)
}
