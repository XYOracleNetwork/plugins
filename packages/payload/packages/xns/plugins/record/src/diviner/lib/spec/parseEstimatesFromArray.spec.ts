import { matchers } from '@xylabs/vitest-matchers'
import type { Payload } from '@xyo-network/payload-model'
import {
  describe, expect, it,
} from 'vitest'

import type { Estimate } from '../parseEstimatesFromArray.ts'
import { parseEstimatesFromArray } from '../parseEstimatesFromArray.ts'

expect.extend(matchers)

describe('parseEstimatesFromArray', () => {
  const estimateA: Estimate = [
    {
      addresses: ['ce5ba87678f8b5e2d0dbccf2fabff65f79191f73'],
      payload_hashes: [
        '49699c9d1cfb1a8643037c9557123a9e643a3dd938f619af39b94966ecfdd6ff',
        'af31f9d6afd8f7cbbe5c3310f2a2da9e1af31d1b7cf8dae7d9786848de1e5bd7',
      ],
      payload_schemas: [
        'network.xyo.hash.lease.estimate',
        'network.xyo.ns.domain.registration.lease',
      ],
      previous_hashes: [null],
      schema: 'network.xyo.boundwitness',
      $signatures: [
        '2a59fea264bf71603f9908ac557cf5816ddd1898ce26a2cbb8081195cf5a19381bf42cc2e583a9a2a768dc324c6ac1344eb4ec06f41b19b06970734b30340226',
      ],
    },
    {
      schema: 'network.xyo.hash.lease.estimate',
      currency: 'USD',
      exp: 1_725_225_427_392,
      nbf: 1_724_966_227_392,
      price: 16,
      $sources: ['dcc7ab18c86d90311bb1b59361f6698e2714294a91aa3126c554aa8483a7ae37'],
    },
    {
      schema: 'network.xyo.ns.domain.registration.lease',
      domain: 'test8446e171-46ca-47d5-a3b7-1a658cb7f451',
      exp: 1_756_502_227_392,
      nbf: 1_724_966_227_392,
      registrant: ['1b2c45910fb0d32cb1d1cd010cbc594bf05969b5'],
      registrar: ['846c35ce6ec62e4608a1215e14b5b16471904153'],
      $sources: ['fccd06633e70c2168c8eb0eb5e25b22e34c1c417a46d5073034ec977dfc1995d', 'd7f40698fb0b0261834ef694ed275b4e6547e329cedc1c56708f098a5608b93a'],
      tld: 'xyo',
    },
  ]
  const estimateB: Estimate = [
    {
      addresses: ['9146761e29463d783e28161848feeeb43196227b'],
      payload_hashes: [
        'abb7ecb2cfc51e1a95f85477b3a4210a72886d5b22fcf32043736c48e2873f2c',
        '3ed825d48507b899395ca943e1dd6ec9a7c948e1cd1573a6cfda66de83af4b11',
      ],
      payload_schemas: [
        'network.xyo.hash.lease.estimate',
        'network.xyo.ns.domain.registration.lease',
      ],
      previous_hashes: [null],
      schema: 'network.xyo.boundwitness',
      $signatures: [
        '91519424f721376751b19865433ff80e3d47b2a59be6f7478e7e003e0366ccc730c1dd7eeaa642ced9e78a40c5a5a3ab33044a6d620f5a4467dfbd0e2d9dc4d8',
      ],
    },
    {
      schema: 'network.xyo.hash.lease.estimate',
      currency: 'USD',
      exp: 1_725_225_514_773,
      nbf: 1_724_966_314_773,
      price: 16,
      $sources: ['90259bc9ecde9d4c0061bc92dbfcf32e0f7b2603e49443d9f3c67a6d7e82c7d5'],
    },
    {
      schema: 'network.xyo.ns.domain.registration.lease',
      domain: 'testb88a0ee2-9665-48b3-b9c6-047eadc475ee',
      exp: 1_756_502_314_773,
      nbf: 1_724_966_314_773,
      registrant: ['1b2c45910fb0d32cb1d1cd010cbc594bf05969b5'],
      registrar: ['846c35ce6ec62e4608a1215e14b5b16471904153'],
      $sources: ['407ed63c0c36fadae7ae8d16a42b678551e0ea6aa5357b76f0ffc1bce3ba2f22', 'ced04c0a3f65bbefde58ba0d15767fcee20b6e25d3820f90c55f44f18effa70f'],
      tld: 'xyo',
    },
  ]
  describe('with valid symmetric data', () => {
    const cases: Payload[][] = [
      [...estimateA],
      [...estimateB],
      [...estimateA, ...estimateB],
      [...estimateB, ...estimateA],
    ]
    it.each(cases)('parses estimates from array', async (...data) => {
      const parsed = await parseEstimatesFromArray(data)
      expect(parsed).toMatchSnapshot()
    })
  })
  describe('with valid jagged data', () => {
    const cases: Payload[][] = [
      estimateA.slice(0, -1),
      estimateB.slice(0, -1),
      [...estimateB],
      [...estimateA, ...estimateB.slice(0, -1)],
      [...estimateB, ...estimateA.slice(0, -1)],
    ]
    it.each(cases)('parses estimates from array', async (...data) => {
      const parsed = await parseEstimatesFromArray(data)
      expect(parsed).toMatchSnapshot()
    })
  })
  describe('with no data', () => {
    it('returns empty array', async () => {
      const parsed = await parseEstimatesFromArray([])
      expect(parsed).toBeArrayOfSize(0)
    })
  })
})
