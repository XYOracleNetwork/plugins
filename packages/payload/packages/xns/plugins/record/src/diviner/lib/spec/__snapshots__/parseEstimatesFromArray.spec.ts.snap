// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseEstimatesFromArray > with valid data > parses estimates from array 1`] = `
[
  [
    {
      "$hash": "b28d75062decc15271a5bc349622662c09f42dbe581a86e09c353186c63bb228",
      "$meta": {
        "signatures": [
          "6d1f0d8d175f0bce32e5e474d263b0c6a815b698563a2d761464acea411bace051fe0fe720905688df8debfb13eed2979fd22f39d07922f06c8f5bebdafa5bb1",
        ],
        "sourceQuery": "d5cd65081af158789b481415644549d72cb6c5925ba68918c4ac55e92bf1f0e2",
        "timestamp": 1724966227501,
      },
      "addresses": [
        "2486705072e91a1d80d97cdcccd5db47b584e4bd",
      ],
      "payload_hashes": [
        "dcc7ab18c86d90311bb1b59361f6698e2714294a91aa3126c554aa8483a7ae37",
        "06882b9889a7649d7172e96c2c5046d5472685cf893c2e38dacd7ef154d7808d",
      ],
      "payload_schemas": [
        "network.xyo.ns.domain.registration.lease",
        "network.xyo.hash.lease.estimate",
      ],
      "previous_hashes": [
        null,
      ],
      "schema": "network.xyo.boundwitness",
      "timestamp": 1724966227501,
    },
    {
      "$hash": "06882b9889a7649d7172e96c2c5046d5472685cf893c2e38dacd7ef154d7808d",
      "currency": "USD",
      "exp": 1725225427392,
      "nbf": 1724966227392,
      "price": 16,
      "schema": "network.xyo.hash.lease.estimate",
      "sources": [
        "dcc7ab18c86d90311bb1b59361f6698e2714294a91aa3126c554aa8483a7ae37",
      ],
    },
    {
      "$hash": "dcc7ab18c86d90311bb1b59361f6698e2714294a91aa3126c554aa8483a7ae37",
      "domain": "test8446e171-46ca-47d5-a3b7-1a658cb7f451",
      "exp": 1756502227392,
      "nbf": 1724966227392,
      "registrant": [
        "1b2c45910fb0d32cb1d1cd010cbc594bf05969b5",
      ],
      "registrar": [
        "846c35ce6ec62e4608a1215e14b5b16471904153",
      ],
      "schema": "network.xyo.ns.domain.registration.lease",
      "sources": [
        "fccd06633e70c2168c8eb0eb5e25b22e34c1c417a46d5073034ec977dfc1995d",
        "d7f40698fb0b0261834ef694ed275b4e6547e329cedc1c56708f098a5608b93a",
      ],
      "tld": "xyo",
    },
  ],
  [
    {
      "$hash": "9366fc2d6e2aa25564077be29088701805a77c096520f7dafe88510eb1fbd4b5",
      "$meta": {
        "signatures": [
          "ec17e78d94d256a0e4e9877855d92ed7bca7f7685ac42b5df87e6395331726a24f6a9271f3f058e1d35e06a11022ff359ccddbfa243772a8e45addad967afb6f",
        ],
        "sourceQuery": "19b2b1e4e0e50747f2cb0134e61cda9523ff9f963a40f3fc9e8e9530834b360b",
        "timestamp": 1724966314885,
      },
      "addresses": [
        "2486705072e91a1d80d97cdcccd5db47b584e4bd",
      ],
      "payload_hashes": [
        "90259bc9ecde9d4c0061bc92dbfcf32e0f7b2603e49443d9f3c67a6d7e82c7d5",
        "7037fc16e12c7cffec09ed1d2ef3f65737668864be38f4a1268026d17601b5ca",
      ],
      "payload_schemas": [
        "network.xyo.ns.domain.registration.lease",
        "network.xyo.hash.lease.estimate",
      ],
      "previous_hashes": [
        null,
      ],
      "schema": "network.xyo.boundwitness",
      "timestamp": 1724966314884,
    },
    {
      "$hash": "7037fc16e12c7cffec09ed1d2ef3f65737668864be38f4a1268026d17601b5ca",
      "currency": "USD",
      "exp": 1725225514773,
      "nbf": 1724966314773,
      "price": 16,
      "schema": "network.xyo.hash.lease.estimate",
      "sources": [
        "90259bc9ecde9d4c0061bc92dbfcf32e0f7b2603e49443d9f3c67a6d7e82c7d5",
      ],
    },
    {
      "$hash": "90259bc9ecde9d4c0061bc92dbfcf32e0f7b2603e49443d9f3c67a6d7e82c7d5",
      "domain": "testb88a0ee2-9665-48b3-b9c6-047eadc475ee",
      "exp": 1756502314773,
      "nbf": 1724966314773,
      "registrant": [
        "1b2c45910fb0d32cb1d1cd010cbc594bf05969b5",
      ],
      "registrar": [
        "846c35ce6ec62e4608a1215e14b5b16471904153",
      ],
      "schema": "network.xyo.ns.domain.registration.lease",
      "sources": [
        "407ed63c0c36fadae7ae8d16a42b678551e0ea6aa5357b76f0ffc1bce3ba2f22",
        "ced04c0a3f65bbefde58ba0d15767fcee20b6e25d3820f90c55f44f18effa70f",
      ],
      "tld": "xyo",
    },
  ],
]
`;
